core/routing_dtlsr.go

```diff
@@ -0,0 +1,750 @@
+		log.WithFields(log.Fields{
+			"NodeID": id,
+		}).Debug("Node already tracked")
+		// node is already tracked
+		return
+	}
+
+	dtlsr.nodeIndex[id] = dtlsr.length
+	dtlsr.indexNode = append(dtlsr.indexNode, id)
+	dtlsr.length = dtlsr.length + 1
+	log.WithFields(log.Fields{
+		"NodeID": id,
+	}).Debug("Added node to tracking store")
+}
+
+// computeRoutingTable finds shortest paths using dijkstra's algorithm
+func (dtlsr *DTLSR) computeRoutingTable() {
+	log.Debug("Recomputing routing table")
+
+	currentTime := timestampNow()
+	graph := dijkstra.NewGraph()
+
+	// add vertices
+	for i := 0; i < dtlsr.length; i++ {
+		graph.AddVertex(i)
+		// log node-index mapping for debug purposes
+		log.WithFields(log.Fields{
+			"index": i,
+			"node":  dtlsr.indexNode[i],
+		}).Debug("Node-index-mapping")
+	}
+
+	// add edges originating from this node
+	for peer, timestamp := range dtlsr.peers.peers {
+		var edgeCost int64
+		if timestamp == 0 {
+			edgeCost = 0
+		} else {
+			edgeCost = int64(currentTime - timestamp)
+		}
+
+		if err := graph.AddArc(0, dtlsr.nodeIndex[peer], edgeCost); err != nil {
+			log.WithFields(log.Fields{
+				"reason": err.Error(),
+			}).Warn("Error computing routing table")
+			return
+		}
+
+		log.WithFields(log.Fields{
+			"peerA": dtlsr.c.NodeId,
+			"peerB": peer,
+			"cost":  edgeCost,
+		}).Debug("Added vertex")
+	}
+
+	// add edges originating from other nodes
+	for _, data := range dtlsr.receivedData {
+		for peer, timestamp := range data.peers {
+			var edgeCost int64
+			if timestamp == 0 {
+				edgeCost = 0
+			} else {
+				edgeCost = int64(currentTime - timestamp)
+			}
+
+			if err := graph.AddArc(dtlsr.nodeIndex[data.id], dtlsr.nodeIndex[peer], edgeCost); err != nil {
+				log.WithFields(log.Fields{
+					"reason": err.Error(),
+				}).Warn("Error computing routing table")
+				return
+			}
+
+			log.WithFields(log.Fields{
+				"peerA": data.id,
+				"peerB": peer,
+				"cost":  edgeCost,
+			}).Debug("Added vertex")
+		}
+	}
+
+	routingTable := make(map[bundle.EndpointID]bundle.EndpointID)
+	for i := 1; i < dtlsr.length; i++ {
+		shortest, err := graph.Shortest(0, i)
+		if err == nil {
+			if len(shortest.Path) <= 1 {
+				log.WithFields(log.Fields{
+					"node_index": i,
+					"node":       dtlsr.indexNode[i],
+					"path":       shortest.Path,
+				}).Warn("Single step path found - this should not happen")
+				continue
+			}
+
+			routingTable[dtlsr.indexNode[i]] = dtlsr.indexNode[shortest.Path[1]]
+			log.WithFields(log.Fields{
+				"node_index": i,
+				"node":       dtlsr.indexNode[i],
+				"path":       shortest.Path,
+				"next_hop":   routingTable[dtlsr.indexNode[i]],
+			}).Debug("Found path to node")
+		} else {
+			log.WithFields(log.Fields{
+				"node_index": i,
+				"error":      err.Error(),
+			}).Debug("Did not find path to node")
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"routingTable": routingTable,
+	}).Debug("Finished routing table computation")
+
+	dtlsr.routingTable = routingTable
+}
```

Comment:
Out of interest I would like to know why Dijkstra is completely recalculated every time. Would it not be possible to perform a partial recalculation for new and disappearing nodes.

Reply:
I don't think, that the building of the graph has any significant performance impact. By far the most expansive part of Dijkstra is the actual shortest-path-computation.